<a href="https://colab.research.google.com/github/lovishkanther11/Deep-Learning/blob/master/KerasInitializations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Usage of initializers

Initializations define the way to set the initial random weights of Keras layers.

The keyword arguments used for passing initializers to layers will depend on the layer. Usually it is simply kernel_initializer and bias_initializer:

Lets test the various initializations of weights on our baseline model.

# Baseline model for classification on CIFAR10 dataset

In [0]:
#import packages
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072, )))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:45: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 100us/step - loss: 1.8911 - acc: 0.3180 - val_loss: 1.7376 - val_acc: 0.3732
Epoch 2/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.6823 - acc: 0.3972 - val_loss: 1.6018 - val_acc: 0.4304
Epoch 3/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5909 - acc: 0.4303 - val_loss: 1.5875 - val_acc: 0.4371
Epoch 4/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5178 - acc: 0.4566 - val_loss: 1.5174 - val_acc: 0.4667
Epoch 5/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.4716 - acc: 0.4753 - val_loss: 1.4804 - val_acc: 0.4754


# Model with Initializer as random uniform and zero

In [0]:
#import packages
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer='random_uniform', bias_initializer='zeros'))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer='random_uniform', bias_initializer='zeros'))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer='random_uniform', bias_initializer='zeros'))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer='random_uniform', bias_initializer='zeros'))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:45: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 100us/step - loss: 1.8565 - acc: 0.3220 - val_loss: 1.7185 - val_acc: 0.3761
Epoch 2/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.6679 - acc: 0.3990 - val_loss: 1.6046 - val_acc: 0.4175
Epoch 3/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.5799 - acc: 0.4320 - val_loss: 1.5674 - val_acc: 0.4384
Epoch 4/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5105 - acc: 0.4575 - val_loss: 1.5206 - val_acc: 0.4469
Epoch 5/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.4615 - acc: 0.4775 - val_loss: 1.4880 - val_acc: 0.4685


#Model with weights as ones

In [0]:
#import packages
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer='ones', bias_initializer='ones'))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer='ones', bias_initializer='ones'))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer='ones', bias_initializer='ones'))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer='ones', bias_initializer='ones'))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:45: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 104us/step - loss: 14.4652 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 2/5
50000/50000 [==============================] - 4s 87us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 3/5
50000/50000 [==============================] - 4s 88us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 4/5
50000/50000 [==============================] - 4s 87us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 5/5
50000/50000 [==============================] - 4s 86us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000


# Model initialized with constant values

Initializer that generates tensors initialized to a constant value.

Arguments

1. value: float; the value of the generator tensors.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.Constant(value=0)
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 105us/step - loss: 2.3027 - acc: 0.0972 - val_loss: 2.3026 - val_acc: 0.1000
Epoch 2/5
50000/50000 [==============================] - 4s 87us/step - loss: 2.3027 - acc: 0.0964 - val_loss: 2.3026 - val_acc: 0.1000
Epoch 3/5
50000/50000 [==============================] - 4s 87us/step - loss: 2.3027 - acc: 0.0986 - val_loss: 2.3026 - val_acc: 0.1000
Epoch 4/5
50000/50000 [==============================] - 4s 86us/step - loss: 2.3027 - acc: 0.0979 - val_loss: 2.3026 - val_acc: 0.1000
Epoch 5/5
50000/50000 [==============================] - 4s 87us/step - loss: 2.3027 - acc: 0.0983 - val_loss: 2.3026 - val_acc: 0.1000


#RandomNormal
keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None)
Initializer that generates tensors with a normal distribution.

Arguments

1. mean: a python scalar or a scalar tensor. Mean of the random values to generate.
2. stddev: a python scalar or a scalar tensor. Standard deviation of the random values to generate.
3. seed: A Python integer. Used to seed the random generator.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None)
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 107us/step - loss: 1.9416 - acc: 0.3177 - val_loss: 1.7641 - val_acc: 0.3738
Epoch 2/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.6620 - acc: 0.4039 - val_loss: 1.5896 - val_acc: 0.4296
Epoch 3/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.5536 - acc: 0.4434 - val_loss: 1.5582 - val_acc: 0.4424
Epoch 4/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5015 - acc: 0.4631 - val_loss: 1.5225 - val_acc: 0.4546
Epoch 5/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.4483 - acc: 0.4805 - val_loss: 1.5060 - val_acc: 0.4634


#RandomUniform
keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None)
Initializer that generates tensors with a uniform distribution.

Arguments

1. minval: A python scalar or a scalar tensor. Lower bound of the range of random values to generate.
2. maxval: A python scalar or a scalar tensor. Upper bound of the range of random values to generate. Defaults to 1 for float types.
3. seed: A Python integer. Used to seed the random generator.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.RandomUniform(minval=-0.05, maxval=0.05, seed=None)
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 107us/step - loss: 1.8622 - acc: 0.3193 - val_loss: 1.7355 - val_acc: 0.3694
Epoch 2/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.6669 - acc: 0.4014 - val_loss: 1.6029 - val_acc: 0.4178
Epoch 3/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5789 - acc: 0.4326 - val_loss: 1.5369 - val_acc: 0.4494
Epoch 4/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.5176 - acc: 0.4551 - val_loss: 1.4976 - val_acc: 0.4744
Epoch 5/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.4603 - acc: 0.4763 - val_loss: 1.4485 - val_acc: 0.4856


#TruncatedNormal
keras.initializers.TruncatedNormal(mean=0.0, stddev=0.05, seed=None)
Initializer that generates a truncated normal distribution.

These values are similar to values from a RandomNormal except that values more than two standard deviations from the mean are discarded and redrawn. This is the recommended initializer for neural network weights and filters.

Arguments

1. mean: a python scalar or a scalar tensor. Mean of the random values to generate.
2. stddev: a python scalar or a scalar tensor. Standard deviation of the random values to generate.
3. seed: A Python integer. Used to seed the random generator.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.TruncatedNormal(mean=0.0, stddev=0.05, seed=None)
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 5s 107us/step - loss: 1.8731 - acc: 0.3266 - val_loss: 1.6885 - val_acc: 0.3911
Epoch 2/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.6625 - acc: 0.4040 - val_loss: 1.6406 - val_acc: 0.4140
Epoch 3/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5644 - acc: 0.4418 - val_loss: 1.5425 - val_acc: 0.4461
Epoch 4/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5087 - acc: 0.4616 - val_loss: 1.5054 - val_acc: 0.4639
Epoch 5/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.4579 - acc: 0.4792 - val_loss: 1.4590 - val_acc: 0.4796


#VarianceScaling
1. keras.initializers.VarianceScaling(scale=1.0, mode='fan_in', distribution='normal', seed=None)
2. Initializer capable of adapting its scale to the shape of weights.

With distribution="normal", samples are drawn from a truncated normal distribution centered on zero, with stddev = sqrt(scale / n) where n is:

1. number of input units in the weight tensor, if mode = "fan_in"
2. number of output units, if mode = "fan_out"
3. average of the numbers of input and output units, if mode = "fan_avg"

With distribution="uniform", samples are drawn from a uniform distribution within [-limit, limit], with limit = sqrt(3 * scale / n).

Arguments

1. scale: Scaling factor (positive float).
2. mode: One of "fan_in", "fan_out", "fan_avg".
3. distribution: Random distribution to use. One of "normal", "uniform".
4. seed: A Python integer. Used to seed the random generator.

Raises

ValueError: In case of an invalid value for the "scale", mode" or "distribution" arguments.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.VarianceScaling(scale=1.0, mode='fan_in', distribution='normal', seed=None)
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 110us/step - loss: 1.8676 - acc: 0.3219 - val_loss: 1.7144 - val_acc: 0.3806
Epoch 2/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.6681 - acc: 0.3990 - val_loss: 1.6174 - val_acc: 0.4148
Epoch 3/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.5683 - acc: 0.4347 - val_loss: 1.5272 - val_acc: 0.4556
Epoch 4/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.5115 - acc: 0.4573 - val_loss: 1.5454 - val_acc: 0.4470
Epoch 5/5
50000/50000 [==============================] - 4s 86us/step - loss: 1.4480 - acc: 0.4808 - val_loss: 1.4659 - val_acc: 0.4777


#Orthogonal
keras.initializers.Orthogonal(gain=1.0, seed=None)
Initializer that generates a random orthogonal matrix.

Arguments

1. gain: Multiplicative factor to apply to the orthogonal matrix.
2. seed: A Python integer. Used to seed the random generator.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.Orthogonal(gain=2.0, seed=None)
model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 113us/step - loss: 14.4010 - acc: 0.1004 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 2/5
50000/50000 [==============================] - 4s 88us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 3/5
50000/50000 [==============================] - 4s 87us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 4/5
50000/50000 [==============================] - 4s 88us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000
Epoch 5/5
50000/50000 [==============================] - 4s 88us/step - loss: 14.5063 - acc: 0.1000 - val_loss: 14.5063 - val_acc: 0.1000


#Identity

1. keras.initializers.Identity(gain=1.0)
2. Initializer that generates the identity matrix.

Only use for 2D matrices. If the desired matrix is not square, it gets padded with zeros for the additional rows/columns.

Arguments

1. gain: Multiplicative factor to apply to the identity matrix.


In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.Identity(gain=1.0)

model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:48: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 126us/step - loss: 2.1912 - acc: 0.1776 - val_loss: 2.0926 - val_acc: 0.2025
Epoch 2/5
50000/50000 [==============================] - 5s 91us/step - loss: 2.0467 - acc: 0.2076 - val_loss: 1.9927 - val_acc: 0.2194
Epoch 3/5
50000/50000 [==============================] - 4s 90us/step - loss: 1.9874 - acc: 0.2219 - val_loss: 1.9508 - val_acc: 0.2418
Epoch 4/5
50000/50000 [==============================] - 5s 91us/step - loss: 1.9531 - acc: 0.2498 - val_loss: 1.9293 - val_acc: 0.2592
Epoch 5/5
50000/50000 [==============================] - 4s 90us/step - loss: 1.9369 - acc: 0.2548 - val_loss: 1.9348 - val_acc: 0.2471


#lecun_uniform
keras.initializers.lecun_uniform(seed=None)

LeCun uniform initializer.

It draws samples from a uniform distribution within [-limit, limit] where limit is sqrt(3 / fan_in) where fan_in is the number of input units in the weight tensor.

Arguments

1. seed: A Python integer. Used to seed the random generator.

Returns

1. An initializer.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.lecun_uniform(seed=None)

model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:48: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 123us/step - loss: 1.8765 - acc: 0.3209 - val_loss: 1.7229 - val_acc: 0.3846
Epoch 2/5
50000/50000 [==============================] - 5s 93us/step - loss: 1.6657 - acc: 0.4042 - val_loss: 1.5895 - val_acc: 0.4383
Epoch 3/5
50000/50000 [==============================] - 5s 92us/step - loss: 1.5605 - acc: 0.4423 - val_loss: 1.5163 - val_acc: 0.4610
Epoch 4/5
50000/50000 [==============================] - 5s 91us/step - loss: 1.5034 - acc: 0.4600 - val_loss: 1.4710 - val_acc: 0.4744
Epoch 5/5
50000/50000 [==============================] - 5s 90us/step - loss: 1.4578 - acc: 0.4773 - val_loss: 1.4432 - val_acc: 0.4874


#glorot_normal

keras.initializers.glorot_normal(seed=None)

Glorot normal initializer, also called Xavier normal initializer.

It draws samples from a truncated normal distribution centered on 0 with stddev = sqrt(2 / (fan_in + fan_out)) where fan_in is the number of input units in the weight tensor and fan_out is the number of output units in the weight tensor.

Arguments

1. seed: A Python integer. Used to seed the random generator.

Returns

1. An initializer.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.glorot_normal(seed=None)

model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:48: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 118us/step - loss: 1.8939 - acc: 0.3135 - val_loss: 1.7741 - val_acc: 0.3691
Epoch 2/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.6610 - acc: 0.4009 - val_loss: 1.5952 - val_acc: 0.4275
Epoch 3/5
50000/50000 [==============================] - 5s 96us/step - loss: 1.5597 - acc: 0.4401 - val_loss: 1.5563 - val_acc: 0.4379
Epoch 4/5
50000/50000 [==============================] - 4s 89us/step - loss: 1.4999 - acc: 0.4627 - val_loss: 1.4770 - val_acc: 0.4690
Epoch 5/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.4491 - acc: 0.4789 - val_loss: 1.4521 - val_acc: 0.4778


#glorot_uniform

keras.initializers.glorot_uniform(seed=None)

Glorot uniform initializer, also called Xavier uniform initializer.

It draws samples from a uniform distribution within [-limit, limit] where limit is sqrt(6 / (fan_in + fan_out)) where fan_in is the number of input units in the weight tensor and fan_out is the number of output units in the weight tensor.

Arguments

1. seed: A Python integer. Used to seed the random generator.

Returns

1. An initializer.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.glorot_uniform(seed=None)


model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:49: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 120us/step - loss: 1.8987 - acc: 0.3213 - val_loss: 1.7977 - val_acc: 0.3301
Epoch 2/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.6706 - acc: 0.3993 - val_loss: 1.5880 - val_acc: 0.4334
Epoch 3/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.5642 - acc: 0.4419 - val_loss: 1.5520 - val_acc: 0.4438
Epoch 4/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.5084 - acc: 0.4601 - val_loss: 1.5223 - val_acc: 0.4447
Epoch 5/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.4558 - acc: 0.4779 - val_loss: 1.4655 - val_acc: 0.4758


#he_normal

keras.initializers.he_normal(seed=None)

He normal initializer.

It draws samples from a truncated normal distribution centered on 0 with stddev = sqrt(2 / fan_in) where fan_in is the number of input units in the weight tensor.

Arguments

1. seed: A Python integer. Used to seed the random generator.

Returns

An initializer.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.he_normal(seed=None)




model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:51: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 121us/step - loss: 2.0272 - acc: 0.3144 - val_loss: 1.7418 - val_acc: 0.3703
Epoch 2/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.6614 - acc: 0.4025 - val_loss: 1.5730 - val_acc: 0.4395
Epoch 3/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.5635 - acc: 0.4405 - val_loss: 1.5963 - val_acc: 0.4291
Epoch 4/5
50000/50000 [==============================] - 4s 89us/step - loss: 1.5037 - acc: 0.4612 - val_loss: 1.4900 - val_acc: 0.4610
Epoch 5/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.4510 - acc: 0.4818 - val_loss: 1.4787 - val_acc: 0.4670


#lecun_normal

keras.initializers.lecun_normal(seed=None)

LeCun normal initializer.

It draws samples from a truncated normal distribution centered on 0 with stddev = sqrt(1 / fan_in) where fan_in is the number of input units in the weight tensor.

Arguments

1. seed: A Python integer. Used to seed the random generator.

Returns

1. An initializer.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.lecun_normal(seed=None)


model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:49: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 123us/step - loss: 1.8766 - acc: 0.3185 - val_loss: 1.7260 - val_acc: 0.3746
Epoch 2/5
50000/50000 [==============================] - 4s 87us/step - loss: 1.6588 - acc: 0.4022 - val_loss: 1.5804 - val_acc: 0.4354
Epoch 3/5
50000/50000 [==============================] - 4s 89us/step - loss: 1.5640 - acc: 0.4390 - val_loss: 1.5525 - val_acc: 0.4399
Epoch 4/5
50000/50000 [==============================] - 4s 90us/step - loss: 1.5016 - acc: 0.4610 - val_loss: 1.4800 - val_acc: 0.4679
Epoch 5/5
50000/50000 [==============================] - 4s 88us/step - loss: 1.4594 - acc: 0.4771 - val_loss: 1.4792 - val_acc: 0.4714


#he_uniform

keras.initializers.he_uniform(seed=None)

He uniform variance scaling initializer.

It draws samples from a uniform distribution within [-limit, limit] where limit is sqrt(6 / fan_in) where fan_in is the number of input units in the weight tensor.

Arguments

1. seed: A Python integer. Used to seed the random generator.

Returns

An initializer.

In [0]:
#import packages
import keras
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.utils import to_categorical

#import dataset
(X_train, y_train), (X_test, y_test) = cifar10.load_data()

#change shape from image to vector
X_train = X_train.reshape(50000, 32 * 32 * 3)
X_test = X_test.reshape(10000, 32 * 32 * 3)

#preprocess
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255.0
X_test /= 255.0

#change labels from numeric to one hot encoded
Y_train = to_categorical(y_train, 10)
Y_test =  to_categorical(y_test, 10)

#Model building
initializer = keras.initializers.he_uniform(seed=None)

model = Sequential()
model.add(Dense(1024, input_shape=(3072,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=5,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:48: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 50000 samples, validate on 10000 samples
Epoch 1/5
50000/50000 [==============================] - 6s 127us/step - loss: 2.0314 - acc: 0.3076 - val_loss: 1.7529 - val_acc: 0.3726
Epoch 2/5
50000/50000 [==============================] - 4s 89us/step - loss: 1.6732 - acc: 0.3994 - val_loss: 1.6260 - val_acc: 0.4217
Epoch 3/5
50000/50000 [==============================] - 4s 89us/step - loss: 1.5753 - acc: 0.4350 - val_loss: 1.5952 - val_acc: 0.4259
Epoch 4/5
50000/50000 [==============================] - 4s 89us/step - loss: 1.5152 - acc: 0.4585 - val_loss: 1.5485 - val_acc: 0.4410
Epoch 5/5
50000/50000 [==============================] - 5s 91us/step - loss: 1.4664 - acc: 0.4752 - val_loss: 1.4809 - val_acc: 0.4613


#Thank you for completing this notebook

**he_normal** has the highest accuracy of 0.4818


**lecun_uniform** has the highest validation accuracy of 0.4874

1)	Baseline model -  acc: 0.4753 , val_acc: 0.4754


2)	Model with Initializer as random uniform and zero - acc: 0.4775, val_acc: 0.4685


3)	Model with weights as one - acc: 0.1000 , val_acc: 0.1000


4)	Model initialized with constant values- acc: 0.0983, val_acc: 0.1000


5)	Random Normal - acc: 0.4805 , val_acc: 0.4634


6)	Random Uniform - acc: 0.4763 , val_acc: 0.4856


7)	Truncated Normal - acc: 0.4792 , val_acc: 0.4796


8)	Variance Scaling -  acc: 0.4808 , val_acc: 0.4777


9)	Orthogonal - acc: 0.1000 , val_acc: 0.1000


10)	Identity - acc: 0.2548 , val_acc: 0.2471


11)	Lecun_uniform - acc: 0.4773 , val_acc: 0.4874


12)	Glorot_normal - acc: 0.4789 , val_acc: 0.4778


13)	Glorot_uniform - acc: 0.4779 , val_acc: 0.4758


14)	He_normal -  acc: 0.4818 , val_acc: 0.4670


15)	Lecun_normal - acc: 0.4771 , val_acc: 0.4714


16)	He_uniform - acc: 0.4752 , val_acc: 0.4613


In [0]:
import keras
from keras import models
from keras.layers import Dense, Dropout
from keras.utils import to_categorical
from keras.datasets import fashion_mnist
from keras.utils.vis_utils import model_to_dot

NUM_ROWS = 28
NUM_COLS = 28
NUM_CLASSES = 10
BATCH_SIZE = 128
EPOCHS = 10

# Load data
(X_train, Y_train), (X_test, Y_test) = fashion_mnist.load_data()



# Reshape data
X_train = X_train.reshape((X_train.shape[0], NUM_ROWS * NUM_COLS))
X_train = X_train.astype('float32') / 255
X_test = X_test.reshape((X_test.shape[0], NUM_ROWS * NUM_COLS))
X_test = X_test.astype('float32') / 255

# Categorically encode labels
Y_train = to_categorical(Y_train, NUM_CLASSES)
Y_test = to_categorical(Y_test, NUM_CLASSES)
#Model building
initializer = keras.initializers.lecun_uniform(seed=None)

model = Sequential()
model.add(Dense(1024, input_shape=(NUM_ROWS * NUM_COLS,), kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(512, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(X_train, Y_train,
                        batch_size=128,
                        nb_epoch=10,
                        verbose=1,
                        validation_data=(X_test, Y_test))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:52: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 60000 samples, validate on 10000 samples
Epoch 1/10
60000/60000 [==============================] - 5s 76us/step - loss: 0.4733 - acc: 0.8267 - val_loss: 0.4032 - val_acc: 0.8530
Epoch 2/10
60000/60000 [==============================] - 4s 68us/step - loss: 0.3486 - acc: 0.8711 - val_loss: 0.3821 - val_acc: 0.8637
Epoch 3/10
60000/60000 [==============================] - 4s 68us/step - loss: 0.3145 - acc: 0.8820 - val_loss: 0.3800 - val_acc: 0.8662
Epoch 4/10
60000/60000 [==============================] - 4s 67us/step - loss: 0.2897 - acc: 0.8922 - val_loss: 0.3671 - val_acc: 0.8656
Epoch 5/10
60000/60000 [==============================] - 4s 67us/step - loss: 0.2704 - acc: 0.8976 - val_loss: 0.3337 - val_acc: 0.8781
Epoch 6/10
60000/60000 [==============================] - 4s 67us/step - loss: 0.2551 - acc: 0.9040 - val_loss: 0.3437 - val_acc: 0.8751
Epoch 7/10
60000/60000 [==============================] - 4s 67us/step - loss: 0.2421 - acc: 0.9084 - val_loss: 0.3172 - val_acc

In [0]:
import pandas
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from keras.utils import np_utils
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

iris_data = load_iris()
	
x = iris_data.data
y_ = iris_data.target.reshape(-1, 1) # Convert data to a single column

# One Hot encode the class labels
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y_)


# Split the data for training and testing
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.01)

initializer = keras.initializers.lecun_uniform(seed=None)


model = Sequential()

model.add(Dense(10, input_shape=(4,),kernel_initializer=initializer, bias_initializer=initializer, activation='relu'))
model.add(Dense(10, kernel_initializer=initializer, bias_initializer=initializer,activation='relu'))
model.add(Dense(3,kernel_initializer=initializer, bias_initializer=initializer, activation='softmax'))

model.compile(loss='categorical_crossentropy',
                  optimizer='adam',
                  metrics=['accuracy'])
    
 

# training
history = model.fit(train_x, train_y,
                        batch_size=5,
                        nb_epoch=10,
                        verbose=2,
                        validation_data=(test_x, test_y))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:47: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.


Train on 148 samples, validate on 2 samples
Epoch 1/10
 - 1s - loss: 1.2996 - acc: 0.3108 - val_loss: 1.8695 - val_acc: 0.0000e+00
Epoch 2/10
 - 0s - loss: 1.1931 - acc: 0.2905 - val_loss: 1.4122 - val_acc: 0.0000e+00
Epoch 3/10
 - 0s - loss: 1.1187 - acc: 0.2973 - val_loss: 1.1475 - val_acc: 0.0000e+00
Epoch 4/10
 - 0s - loss: 1.0553 - acc: 0.4324 - val_loss: 1.0278 - val_acc: 1.0000
Epoch 5/10
 - 0s - loss: 0.9831 - acc: 0.6757 - val_loss: 1.0037 - val_acc: 1.0000
Epoch 6/10
 - 0s - loss: 0.8665 - acc: 0.7568 - val_loss: 0.9491 - val_acc: 0.5000
Epoch 7/10
 - 0s - loss: 0.7457 - acc: 0.7230 - val_loss: 0.9118 - val_acc: 0.5000
Epoch 8/10
 - 0s - loss: 0.6627 - acc: 0.7635 - val_loss: 0.8489 - val_acc: 1.0000
Epoch 9/10
 - 0s - loss: 0.5995 - acc: 0.8108 - val_loss: 0.7369 - val_acc: 1.0000
Epoch 10/10
 - 0s - loss: 0.5567 - acc: 0.8311 - val_loss: 0.7592 - val_acc: 1.0000
